src/metabase/integrations/slack.clj

```diff
@@ -135,84 +144,96 @@
+  "Calls Slack API `users.list` endpoint and returns the list of available users with their @ prefix. Note the call
+  to [[paged-list-request]] will only fetch the first [[max-list-results]] items."
   [& {:as query-parameters}]
-  (->> (paged-list-request "users.list" :members query-parameters)
+  (->> (paged-list-request "users.list"
+                           ;; response -> user names
+                           #(->> %
+                                 :members
+                                 (map (fn [member]
+                                        (str \@ (:name member)))))
+                           query-parameters)
        ;; filter out deleted users and bots. At the time of this writing there's no way to do this in the Slack API
        ;; itself so we need to do it after the fact.
        (filter (complement :deleted))
        (filter (complement :is_bot))))
 
-(def ^:private ^{:arglists '([channel-name])} files-channel*
-  ;; If the channel has successfully been created we can cache the information about it from the API response. We need
-  ;; this information every time we send out a pulse, but making a call to the `conversations.list` endpoint everytime we
-  ;; send a Pulse can result in us seeing 429 (rate limiting) status codes -- see
-  ;; https://github.com/metabase/metabase/issues/8967
-  ;;
-  ;; Of course, if `files-channel*` *fails* (because the channel is not created), this won't get cached; this is what
-  ;; we want -- to remind people to create it
-  ;;
-  ;; The memoized function is paramterized by the channel name so that if the name is changed, the cached channel details
-  ;; will be refetched.
-  (memoize/ttl
-   (fn [channel-name]
-     (or (when channel-name (channel-with-name channel-name))
-         (let [message (str (tru "Slack channel named `{0}` is missing!" channel-name)
-                            " "
-                            (tru "Please create or unarchive the channel in order to complete the Slack integration.")
-                            " "
-                            (tru "The channel is used for storing images that are included in dashboard subscriptions."))]
-           (log/error (u/format-color 'red message))
-           (throw (ex-info message {:status-code 400})))))
-   :ttl/threshold (u/hours->ms 6)))
+(defn refresh-channels-and-usernames!
+  "Refreshes users and conversations in slack-cache. If called with no args, finds both in parallel, sets
+  [[slack-cached-channels-and-usernames]], and resets the [[slack-channels-and-usernames-last-updated]] time."
+  []
+  (when (slack-configured?)
+    (log/info "Refreshing slack channels and usernames.")
+    (let [users (future (vec (users-list)))
+          conversations (future (vec (conversations-list)))]
+      (slack-cached-channels-and-usernames (concat @conversations @users))
+      (slack-channels-and-usernames-last-updated (t/zoned-date-time)))))
```

Comment:
It might be worth considering putting some locking of some sort in here so we only attempt one update at a time for a given instance. If I make 100 calls to `refresh-channels-and-usernames!` and the first one triggers a refresh it would be nice if the other 99 just waited for the results of the first rather than issuing 100 Slack API requests. I would do something like this:

```clj
(defonce ^:private refresh-lock (Object.))

(defn- refresh-channels-and-usernames!* [] 
  ;; actual refresh logic goes here
  ...
  ;; update the last refresh time
  (slack-channels-and-usernames-last-updated (t/zoned-date-time)))

(defn- needs-refresh? []
  (< (slack-channels-and-usernames-last-updated) 10-minutes-ago))

(defn refresh-channels-and-usernames! []
  (when (needs-refresh?)
    (locking refresh-lock
      (when (needs-refresh?)
        (refresh-channels-and-usernames!*)))))
```

Basic idea is every time `refresh-channels-and-usernames!` is called, first double-check that a refresh is needed, by looking at the value of the `slack-channels-and-usernames-last-updated` Setting. If we actually need a refresh, then wait to acquire the `refresh-lock`; once we acquire that, check `slack-channels-and-usernames-last-updated` again in case it has been updated while we were waiting to get the lock. If we still need to refresh then we can do the actual refresh stuff, and update `slack-channels-and-usernames-last-updated`. Otherwise we can no-op.

Does that make sense?

Reply:
I think that should be sufficient -- if we were really sophisticated we could try to avoid doing simultaneous refreshes across multiple instances but I think that's way more trouble than it's worth. As long as we limit ourselves to one refresh at a time per instance I think it's fine -- it will keep us from blowing up instances with tons of requests